pkg/object/scale_view_lock_manager.go

```diff
@@ -0,0 +1,218 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package object
+
+import (
+	"errors"
+	"math"
+	"sort"
+
+	"github.com/samber/lo"
+	"k8s.io/utils/pointer"
+
+	risingwavev1alpha1 "github.com/risingwavelabs/risingwave-operator/apis/risingwave/v1alpha1"
+)
+
+type ScaleViewLockManager struct {
+	risingwave *risingwavev1alpha1.RisingWave
+}
+
+func (svl *ScaleViewLockManager) getScaleViewLockIndex(sv *risingwavev1alpha1.RisingWaveScaleView) (int, *risingwavev1alpha1.RisingWaveScaleViewLock) {
+	scaleViews := svl.risingwave.Status.ScaleViews
+	for i, s := range scaleViews {
+		if s.Name == sv.Name && s.UID == sv.UID {
+			return i, &scaleViews[i]
+		}
+	}
+	return 0, nil
+}
+
+func (svl *ScaleViewLockManager) GetScaleViewLock(sv *risingwavev1alpha1.RisingWaveScaleView) *risingwavev1alpha1.RisingWaveScaleViewLock {
+	_, r := svl.getScaleViewLockIndex(sv)
+	return r
+}
+
+func (svl *ScaleViewLockManager) IsScaleViewLocked(sv *risingwavev1alpha1.RisingWaveScaleView) bool {
+	return svl.GetScaleViewLock(sv) != nil
+}
+
+func canonizeScalePolicy(p risingwavev1alpha1.RisingWaveScaleViewSpecScalePolicy) risingwavev1alpha1.RisingWaveScaleViewSpecScalePolicy {
+	r := p
+	if r.MaxReplicas == nil {
+		r.MaxReplicas = pointer.Int32(math.MaxInt32)
+	}
+	return r
+}
+
+func (svl *ScaleViewLockManager) split(total, n int) int {
+	if total%n == 0 {
+		return total / n
+	} else {
+		return total/n + 1
+	}
+}
+
+func (svl *ScaleViewLockManager) splitReplicasIntoGroups(sv *risingwavev1alpha1.RisingWaveScaleView) map[string]int32 {
+	// Must be a stable algorithm.
+
+	groupsGroupByPriority := make(map[int32][]risingwavev1alpha1.RisingWaveScaleViewSpecScalePolicy)
+	for _, p := range sv.Spec.ScalePolicy {
+		if _, ok := groupsGroupByPriority[p.Priority]; !ok {
+			groupsGroupByPriority[p.Priority] = []risingwavev1alpha1.RisingWaveScaleViewSpecScalePolicy{
+				canonizeScalePolicy(p),
+			}
+		} else {
+			groupsGroupByPriority[p.Priority] = append(groupsGroupByPriority[p.Priority], canonizeScalePolicy(p))
+		}
+	}
+
+	for k := range groupsGroupByPriority {
+		groups := groupsGroupByPriority[k]
+		sort.Slice(groups, func(i, j int) bool {
+			if *groups[i].MaxReplicas != *groups[j].MaxReplicas {
+				return *groups[i].MaxReplicas < *groups[j].MaxReplicas
+			}
+			return groups[i].Group < groups[j].Group
+		})
+	}
+
+	totalLeft := int(sv.Spec.Replicas)
+	replicas := make(map[string]int32)
+
+	for priority := int32(10); priority >= 0; priority-- {
```

Comment:
Very nested 😂 Would it be possible to refactor it into separate methods for better debugging / extensibility

Reply:
Moved it into another package 😄